korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -281,68 +329,313 @@ abstract class BmpSlice(
+        val x = (tl_x * baseWidth).roundToInt()
+        val y = (tl_y * baseHeight).roundToInt()
+        if (tl_x == tr_x) {
+            intArrayOf(
+                if (tl_x > br_x) x - 1 else x,
+                if (tl_y > br_y) y - 1 else y,
+                0, if (tl_x < br_x) 1 else -1,
+                0, if (tl_y < br_y) 1 else -1
+            )
+        } else {
+            intArrayOf(
+                if (tl_x > br_x) x - 1 else x,
+                if (tl_y > br_y) y - 1 else y,
+                if (tl_x < br_x) 1 else -1, 0,
+                if (tl_y < br_y) 1 else -1, 0
+            )
+        }
+    }
+
+    fun isValidBasePixelPos(x: Int, y: Int): Boolean = x in 0 until frameWidth && y in 0 until frameHeight
+
+    fun basePixelPos(x: Int, y: Int, out: PointInt = PointInt()): PointInt? = if (isValidBasePixelPos(x, y)) basePixelPosUnsafe(x, y, out) else throw IllegalArgumentException("Point $x,$y is not in bounds of slice")
+
+    fun basePixelPosUnsafe(x: Int, y: Int, out: PointInt = PointInt()): PointInt? = out.also {
+        val offsetX = x - frameOffsetX
+        val offsetY = y - frameOffsetY
+        if (offsetX < 0 || offsetY < 0 || offsetX >= width || offsetY >= height)
+            return null
+
+        it.x = pixelOffsets[0] + pixelOffsets[2] * offsetX + pixelOffsets[3] * offsetY
+        it.y = pixelOffsets[1] + pixelOffsets[4] * offsetY + pixelOffsets[5] * offsetX
+    }
+
     fun readPixels(x: Int, y: Int, width: Int, height: Int, out: RgbaArray = RgbaArray(width * height), offset: Int = 0): RgbaArray {
-        check(x in 0 until this.width)
-        check(y in 0 until this.height)
-        check((x + width) in 0 .. this.width)
-        check((y + height) in 0 .. this.height)
+        check(isValidBasePixelPos(x, y))
+        check(isValidBasePixelPos(x + width - 1, y + height - 1))
         check(out.size >= offset + width * height)
-        bmpBase.readPixelsUnsafe(left + x, top + y, width, height, out, offset)
+        readPixelsUnsafe(x, y, width, height, out, offset)
         return out
     }
 
-    fun getRgba(x: Int, y: Int): RGBA = bmpBase.getRgba(left + x, top + y)
-    fun setRgba(x: Int, y: Int, value: RGBA): Unit = bmpBase.setRgba(left + x, top + y, value)
+    fun readPixelsUnsafe(x: Int, y: Int, width: Int, height: Int, out: RgbaArray, offset: Int = 0) {
+        var n = offset
+        val p = PointInt()
+        for (y0 in 0 until height) for (x0 in 0 until width) {
+            if (basePixelPosUnsafe(x0 + x, y0 + y, p) != null) {
+                out[n++] = bmpBase.getRgba(p.x, p.y)
+            } else {
+                out[n++] = Colors.TRANSPARENT_BLACK
+            }
+        }
+    }
+
+    fun getRgba(x: Int, y: Int): RGBA {
+        basePixelPos(x, y)?.let {
+            return bmpBase.getRgba(it.x, it.y)
+        }
+        return Colors.TRANSPARENT_BLACK
+    }
+
+    fun setRgba(x: Int, y: Int, value: RGBA) {
+        basePixelPos(x, y).also {
```

Comment:
This seems to be allocating, maybe not acceptable here because this kind of method is going to be executed like a lot of times.

Not sure about the details but what about creating two methods:

basePixelPosX and basePixelPosY, so it can return a simple Int and no allocation is needed? You can still keep basePixelPos calling both of them to keep the method is external consumers can accept that trade-off

Reply:
Found a little problem,..

To calculate the `basePixelPos` I need to know the `x` and `y` (because of the "`if (tl_x == tr_x) {` problem"). So if I have `basePixelPosX` and `basePixelPosY` most of the validation and calculation have to be done twice.

But `basePixelPos` has an optional parameter `out: PointInt = PointInt()`. Do you think it's ok to create a `private tmpPointInt = PointInt()` and let `set/getRgba()` use it / have an optional parameter `out: PointInt = tmpPointInt`, or will this lead to multithreading problems?